project-lifecycle.md

```diff
@@ -0,0 +1,101 @@
+# Protocol Labs Project Lifecycle
+
+* Outline
+  * Active States
+    * L0: Prototype
+    * L1: “In Org”
+    * L2: Contributors Welcome!
+    * L3: Growth
+    * L4: Top Project
+    * L5: Flagship Project
+  * Inactive States
+    * D0: Actively Maintained
+    * D1: Critical Fixes Only
+    * D2: Hard Deprecated
+
+## Lifecycles and API Stability
+
+For the most part, a project’s place in the lifecycle does not describe it’s API stability. 
+API stability is required once enough people depend on that API and can’t always be planned for.
+
+However, it is assumed that once a project reaches L3 the project is actively soliciting 
+users and is dedicated to supporting them by not breaking API unless absolutely necessary. 
+Similarly, API changes should not happen in inactive states as we are no longer soliciting 
+users and are only investing to the extent that people already depend on them.
+
+## L0: Prototype
+
+*Examples: proto.school*
+
+During the initial stages of development projects exist either on developer’s personal GitHub 
+accounts or in specific orgs like shipyard that are intended to house experimental development 
+and demo’s.
+
+At this stage, attention is a liability, and providing a nice website, good documentation, 
+or other promotion would only bring unwanted attention too early.
+
+## L1: “In Org”
+
+*Examples: interface-ipld*
+
+When projects are moving into relevant PL orgs they are expected to have:
+
+* Guaranteed test coverage at a reasonable level.
+* TBD: other automation.
```

Comment:
> Does L1 explicitly indicate that PL is dedicated to supporting the project going forward?

We're supporting it *while it is L1 or higher* but it could still be moved to D0 and we'd move away from it.

> I’m thinking about the confusion brought about by repos like ipfs-mans or ipfs-nodeschool, where there were only a few commits on a single day and no further effort.

Those are probably L0. They look like prototypes we built and then let go of. This will happen a lot, and that's why there is a level for them where they aren't in main orgs or messaged as being supported.

> Also, are there implications on licensing here? What if someone starts something as a personal project in their own account, but it moves to the IPFS org — does the license need to be reassigned to PL, or do we expect that some “in org” projects have individual copyright?

There's two terms we need to be clear about: *copyright ownership* and *copyright license*. 

For the most part, people own their contributions. Companies can own contributions they've paid for but it's actually quite rare that a company will **transfer** ownership from an individual to the company, mostly because certain countries (Germany) just don't recognize a complete transfer of intellectual property. 

A license is the rights non-owners are granted to the property. Public licenses grant the public rights to the work. For the most part, we only need to care that the work is licensed under a good public licenses (which ones we use/recommend is a longer conversation). We don't need to own the copyright. We *may* need to own the trademark if there is one, but that's a totally different legal structure than copyright which would take a long time to discuss :)

> Also also, do we make some guarantees about READMEs and descriptions at this point? (or is that L2?) I’m thinking of ipfs/docs#55.

That's up for discussion. The important thing here is that L1 needs consistent tooling/docs to make it useful *within* the project and L2 needs consistent tooling/docs for it to be useful to people *outside* the project.

Reply:
> We're supporting it while it is L1 or higher but it could still be moved to D0 and we'd move away from it.

Awesome. I feel like it might be worth stating that explicitly, e.g. `* Commitment to support and grow this project going forward and, if that changes by moving to D0/D1/D2, we’ll make that explicitly clear.`

> Those are probably L0. They look like prototypes we built and then let go of. This will happen a lot, and that's why there is a level for them where they aren't in main orgs or messaged as being supported.

Maybe worth noting this as a descriptive point? e.g. `Being a repo in one of the main orgs signals L1+ status, so most projects should not start there. We haven’t done that well in past, but part of the goal with this plan is to change that.`

> There's two terms we need to be clear about: copyright ownership and copyright license.

Yes! Sorry, I should have written “does the *copyright line in the* license need to be reassigned.” But you answered that question anyway :)

> we only need to care that the work is licensed under a good public licenses (which ones we use/recommend is a longer conversation)

I think it is worth making this a clear point in the doc. L1+ projects must have a public, open source (?) license. Seems reasonable not to specify exactly which licenses are ok here — we’ve had long [discussions](https://github.com/ipfs/community/issues/139) and have [a policy](https://github.com/ipfs/community/blob/07b1120f292afdc027e8be5f256ae1b130a506a1/LICENSING_POLICY.md) on what licenses should be used for original PL projects, but agree that projects coming from elsewhere/other people might not use that specific set of licenses, and that’s ok.

Reply:
> Maybe worth noting this as a descriptive point? e.g. Being a repo in one of the main orgs signals L1+ status, so most projects should not start there. We haven’t done that well in past, but part of the goal with this plan is to change that.

Probably a good idea, but I want to hear what @diasdavid has to say about the lifecycle before I sign us up for such a big shift in policy.

> I think it is worth making this a clear point in the doc, e.g. “L1+ projects must have a public, open source (?) license.” Seems reasonable not to specify exactly which licenses are ok here — we’ve had long discussions and have a policy on what licenses should be used for original PL projects, but agree that projects coming from elsewhere/other people might not use that specific set of licenses, and that’s ok.

This probably needs to go in another document, and to some extent I think *is* already documented elsewhere, and we can link here. We also require the DCO and signoff signatures in commits. There's a ton of rules for being "in org" we need to tighten up. For now, I'm going to say that it's out of scope for this doc and find something to link to that we can iterate on later.